openai/client.py

```diff
@@ -0,0 +1,392 @@
+        """Populate default arguments based on the current client configuration/defaults"""
+        kwargs.setdefault("api_base", self.api_base or openai.api_base)
+        kwargs.setdefault("api_key", self.auth.get_token())
+        kwargs.setdefault("api_type", self.api_type)
+        if self.api_version:
+            kwargs.setdefault("api_version", self.api_version)
+
+        for key, val in overrides.items():
+            if val == ...:
+                continue
+            kwargs.setdefault(key, val)
+            if kwargs[key] != val:
+                raise TypeError(f"No parameter named `{key}`")
+
+    def _normalize_model(self, kwargs: typing.Dict[str, typing.Any]):
+        """Normalize model/engine/deployment_id based on which backend the client is
+        configured to target.
+
+        Specifically, it will pass the provided `model` parameter as `deployment_id`
+        unless `deployment_id` is explicitly passed in.
+        """
+        if (
+            len(
+                [
+                    param
+                    for param in kwargs
+                    if param in ("deployment_id", "model", "engine")
+                ]
+            )
+            != 1
+        ):
+            raise TypeError(
+                "You must specify exactly one of `deployment_id`, `model` and `engine`"
+            )
+
+        if self.backend == "azure":
+            try:
+                # We'll try to "rename" the `model` keyword to fit azure's `deployment_id`
+                # paradigm
+                kwargs["deployment_id"] = kwargs.pop("model")
+            except KeyError:
+                pass
+
+    def completion(self, prompt: str, **kwargs) -> openai.Completion:
+        self._populate_args(kwargs, prompt=prompt, stream=False)
+        self._normalize_model(kwargs)
+        return typing.cast(openai.Completion, openai.Completion.create(**kwargs))
+
+    async def acompletion(self, prompt: str, **kwargs) -> openai.Completion:
+        self._populate_args(kwargs, prompt=prompt, stream=False)
+        self._normalize_model(kwargs)
+        return typing.cast(openai.Completion, await openai.Completion.acreate(**kwargs))
+
+    def iter_completion(
+        self, prompt: str, **kwargs
+    ) -> typing.Iterable[openai.Completion]:
+        self._populate_args(kwargs, prompt=prompt, stream=True)
+        self._normalize_model(kwargs)
+        return typing.cast(
+            typing.Iterable[openai.Completion], openai.Completion.create(**kwargs)
+        )
+
+    async def aiter_completion(
+        self, prompt: str, **kwargs
+    ) -> typing.AsyncIterable[openai.Completion]:
+        self._populate_args(kwargs, prompt=prompt, stream=True)
+        self._normalize_model(kwargs)
+        return typing.cast(
+            typing.AsyncIterable[openai.Completion],
+            await openai.Completion.acreate(**kwargs),
+        )
+
+    def chatcompletion(self, messages, **kwargs) -> openai.ChatCompletion:
```

Comment:
nit: since we snake_case the image API names, this kind of feels like it should follow that as well

Reply:
Yeah, I do think we should review the names across the board. I used the convention of `iter_` for server sent event streaming APIs and `_<variation name>` for "child"/custom actions beyond the base generation API. But that was just the first thing that popped into my head.